### PR TITLE
LPS-186211 Source formatter does not recognize feature flags in client-extensions-api

### PR DESCRIPTION
**Pull request by drewbrokke**
This is an update for [LPS-186211](https://issues.liferay.com/browse/LPS-186211).<br/><br/>/cc @bryceosterhaus

### CODE REVIEW

**Comment by drewbrokke** [comment]
ci:forward

**Comment by bryceosterhaus** [comment]
cc @izaera you might want to take a look here

File: modules/apps/client-extension/client-extension-type-impl/src/main/java/com/liferay/client/extension/type/internal/factory/CETImplFactories.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+	public CETImplFactory<?> getCETImplFactory(String key) {
+		if (_isDisabled(key)) {
+			return null;
+		}
+
+		return _cetImplFactories.get(key);
+	}
+
+	public Set<String> getTypes() {
+		SortedSet<String> types = new TreeSet<>(_cetImplFactories.keySet());
```
Comment (izaera) [inline modules/apps/client-extension/client-extension-type-impl/src/main/java/com/liferay/client/extension/type/internal/factory/CETImplFactories.java:83]: Why sorted? Isn't it a waste of CPU time adn RAM? :thinking:

File: modules/apps/client-extension/client-extension-type-impl/src/main/java/com/liferay/client/extension/type/internal/factory/CETImplFactories.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+		types.removeIf(this::_isDisabled);
+
+		return Collections.unmodifiableSortedSet(types);
+	}
+
+	private boolean _isDisabled(String key) {
+		if ((key.equals(ClientExtensionEntryConstants.TYPE_FDS_CELL_RENDERER) &&
```
Comment (izaera) [inline modules/apps/client-extension/client-extension-type-impl/src/main/java/com/liferay/client/extension/type/internal/factory/CETImplFactories.java:91]: I assume this approach is necessary for the SF patterns to work, but it's penalizing performance at runtime in favor of some build time process, which doesn't seem a good idea.

Imagine if we had 500 types of CXs (which may happen in the long term): every time you requested a factory this code would have to execute an average of 250 comparisons :roll_eyes:

File: modules/apps/client-extension/client-extension-type-impl/src/main/java/com/liferay/client/extension/type/internal/factory/CETImplFactories.java (hunk 1 above)
Comment (izaera) [reply modules/apps/client-extension/client-extension-type-impl/src/main/java/com/liferay/client/extension/type/internal/factory/CETImplFactories.java:83]: Also, in the previous code `_type` was set once and reused all the time which made it unnecessary to create new objects every time `getTypes()` is invoked.

File: modules/apps/client-extension/client-extension-type-impl/src/main/java/com/liferay/client/extension/type/internal/factory/CETImplFactories.java (hunk 2 above)
Comment (izaera) [reply modules/apps/client-extension/client-extension-type-impl/src/main/java/com/liferay/client/extension/type/internal/factory/CETImplFactories.java:91]: Maybe if we leave the previous code and just add a commentary like:

```
// FeatureFlagManagerUtil.isEnabled("LPS-172904")
```

on top of each map value, that would be enough for SF to detect.

**Comment by drewbrokke** [comment]
@izaera I generally agree that this change _shouldn't_ be necessary at all, so I won't fight too hard on it. But I am curious which part alarms you the most - is the primary concern that we are moving from a map (hash comparison) to direct string comparison?

File: modules/apps/client-extension/client-extension-type-impl/src/main/java/com/liferay/client/extension/type/internal/factory/CETImplFactories.java (hunk 2 above)
Comment (izaera) [reply modules/apps/client-extension/client-extension-type-impl/src/main/java/com/liferay/client/extension/type/internal/factory/CETImplFactories.java:91]: Yes, but if you have 500 CX types, you will have 500 chained ifs which, in average, will lead to 250 comparisons. Am I wrong? :thinking:

File: modules/apps/client-extension/client-extension-type-impl/src/main/java/com/liferay/client/extension/type/internal/factory/CETImplFactories.java (hunk 1 above)
Comment (izaera) [reply modules/apps/client-extension/client-extension-type-impl/src/main/java/com/liferay/client/extension/type/internal/factory/CETImplFactories.java:83]: Oh yes, you are right. It was sorted [here](https://github.com/drewbrokke/liferay-portal/pull/1445/files#diff-01abd0524cce98e01755d244f6e13bb7b8faa0d7d4506ee1c6ce4ff2b787d0e6L95)...

So, this code is even worse than I thought :sweat_smile: .

Where we had a [constant time operation with no RAM comsumption](https://github.com/drewbrokke/liferay-portal/pull/1445/files#diff-01abd0524cce98e01755d244f6e13bb7b8faa0d7d4506ee1c6ce4ff2b787d0e6L153) now we have to:

    1. Clone a set just to return it (more RAM)
    2. Sort the cloned set
    3. Filter it
    
Also, remember that the filtering is invoking `isDisabled` which has the chained ifs. Thus, if you had 500 CX types, you need to execute 500 * 250 ifs on average :grimacing:

**Comment by izaera** [comment]
> is the primary concern that we are moving from a map (hash comparison) to direct string comparison?

@drewbrokke yes, mostly that. I find it sad that we make production code worse because we want to please a build tool, as I said... :shrug:

**Comment by drewbrokke** [comment]
@izaera agree.

I like your idea of a comment to "trick" SF, so if Brian will accept that I'd say let's go for that. 

However, I do think we should still remove the `FEATURE_FLAGS` map from the public API - I think we can make the map internal and still maintain the status quo in terms of performance. Thoughts?

**Comment by izaera** [comment]
> However, I do think we should still remove the FEATURE_FLAGS map from the public API - I think we can make the map internal and still maintain the status quo in terms of performance. Thoughts?

IINM, it was internal but someone (Greg?) took it out to the API. No idea why...

I agree that it is better hidden in the impl since having it in the API serves no purpose (apparently) and means that we must update the baseline each time FFs change. :+1:

**Comment by bryceosterhaus** [comment]
@drewbrokke whats the status on this? Just need to figure out when these FF will make it into a release

**Comment by drewbrokke** [comment]
See https://github.com/drewbrokke/liferay-portal/pull/1449

**Pull request closed without merging**